### ISSUE TITLE
Prefix <T> to access methods with generics : `Klass[<T>"method"]`

### ISSUE THREAD

**Issue by denis-migdal** [microsoft/TypeScript#57297]
### 🔍 Search Terms

generic methods

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

Use prefix `<T>` to access methods with generics:
```typescript
class Klass {
    method<T>() {}
}

type Method = Klass[<number>"method"] // <number>() => void
```

Could also be extended for all `<T>(...args: any[]) => any` types :
```typescript
type Fct = <T>() => void;
type FctT<U> = <U>Fct; // FctT<U> = <U>() => void;
```

I think there are no ambiguity with this syntax:
```typescript
typeA<T>typeB  // syntax error
<T>typeA.typeB // = (<T>typeA).typeB
typeA.<T>typeB // = <T>(typeA.typeB)
```

### 📃 Motivating Example

Currently, there are no ways to access a method with generics from a class.

This would solve numerous issues :

- https://github.com/microsoft/TypeScript/issues/57102
- https://github.com/microsoft/TypeScript/issues/50481
- https://github.com/microsoft/TypeScript/pull/47607#issuecomment-1058192563
- https://github.com/microsoft/TypeScript/issues/53410

### 💻 Use Cases

1. What do you want to use this for?

Access methods with generics

3. What shortcomings exist with current approaches?

No current approaches.

4. What workarounds are you using in the meantime?

- https://github.com/microsoft/TypeScript/issues/57102#issuecomment-1902131889 : won't work if the class isn't known
- https://github.com/microsoft/TypeScript/issues/50481#issuecomment-1925327665 : requires to register in an interface all classes that will be used.

**Comment by whzx5byb** [comment]
Workaround: access method from `prototype`.

```ts
type Method = typeof Klass.prototype.method<number>;
```

**Comment by denis-migdal** [comment]
> Workaround: access method from `prototype`.
> 
> ```ts
> type Method = typeof Klass.prototype.method<number>;
> ```

Nope.
You didn't "access a method with generics from a class" in TS, but asked the type of a value that happens to be a method.

```typescript
class Klass {
	method<T>() {

	}
}

// from the class
type T = Klass;
type m1 = T["method"]

// from the type of the class
type TK = typeof Klass;
type m2 = TK["prototype"]["method"];
```

You solution could work using functions :
```typescript
function _getMethod<T extends BaseClass, U>(t: typeof T) {
      return t.prototype.method<U>;
}

type getMethod<T extends BaseClass, U> = ReturnType<typeof _getMethod<T, U>>
```
But due to Typescript widening, that will return `BaseClass.prototype.method<U>` instead of `T.prototype.method<U>`

cf https://github.com/microsoft/TypeScript/issues/57286

Also, we could help TS by precising the return type :
```typescript
function _getMethod<T extends BaseClass, U>(t: typeof T): T["prototype"]["method"] {
      return t.prototype.method<U>;
}
```

But we have currently no ways to precise the generic parameters to the type...

**Comment by RyanCavanaugh** [comment]
> Nope. You didn't "access a method with generics from a class" in TS, but asked the type of a value that happens to be a method.

I don't understand what you're getting at here. @whzx5byb's answer isn't even a workaround; it's a solution.

I also don't see what this being in a class has to do with anything. There are existing issues around generic instantiation of a type as opposed to a value.

**Comment by denis-migdal** [comment]
> I also don't see what this being in a class has to do with anything. There are existing issues around generic instantiation of a type as opposed to a value.

My suggestion wasn't about generic instantiation of a type, but to be able to fetch a generic version of a method.
It is not (2 operations) :
1. operator `[]`: we fetch `Klass["method"]`
2. operator `<>`: we make a generic instantiation of the returned type.

But (1 operation):
1. operator `[<> ]`: we fetch `Klass[<T>"method"]` directly as a type.


I then suggested an extension to the syntax for a more generic usage, that is indeed a generic instantiation of a type.
Which you give (in the same case as above): `<T>(Klass["method"])`. So it is slightly different.

I made another issue as other issues doesn't seem active / followed.
I think I suggested a way that seems to not have the drawback of other cited suggestions:
- no syntax ambiguity (?)
- simple, clear and concise syntax.
- I guess not too much trouble to implement (?).

> > Nope. You didn't "access a method with generics from a class" in TS, but asked the type of a value that happens to be a method.
>
> I don't understand what you're getting at here. @whzx5byb's answer isn't even a workaround; it's a solution.

It isn't, as he gets the type from a **_value_** that happens to be a method, not from the class in TS.

When manipulating types, there are lot of cases where you don't have access to the values :

```typescript
function foo<T extends Record<string, X>>(...) {

     type T2<U> = {
         [K in keyof T]: T[K][<U>"method"] // can't have access to the class value as we are in TS (type manipulation, not value manipulation).
     }
}
```
